### ISSUE TITLE
[DISCUSS] Failsafe Roadmap

### ISSUE THREAD

**Issue by jhalterman** [jhalterman/failsafe#159]
I wanted to sketch out few things I've been working on and considering for upcoming roadmap for Failsafe:

For 2.0:

- [x] Policy SPI, so users can supply 3rd party policies
- [x] Composable policy execution
- [ ] Re-thinking event listeners - See #154
- [ ] More unified execution implementation

Beyond 2.0:

- [ ] More listeners to facilitate metrics integration
- [ ] More Policy implementations including Timeout, Bulkhead and Cache

**Comment by whiskeysierra** [comment]
> * More Policy implementations including Bulkhead, RateLimiter, and Cache

What about *backup requests* or *hedged requests*?
- https://blog.acolyer.org/2015/01/15/the-tail-at-scale/
- https://github.com/zalando/riptide/tree/master/riptide-backup

**Comment by jhalterman** [comment]
@whiskeysierra Do you have another pointer for backup requests?

I suspect both of these would be specific to Http or other client/server technologies. Do you think there's a way to write policies for these in a way that's not specific to particular technologies, and in particular, external dependencies?

**Comment by whiskeysierra** [comment]
Technically it's just a different flavor of retries. Retries right now are triggered by exceptions (or unacceptable results) while backup requests are somewhat like an *early retry* caused by latency/delay. Can apply to anything, not just HTTP.

**Comment by jhalterman** [comment]
Thanks - so to be clear, Riptide's backup requests are the same pattern as hedged requests?

**Comment by whiskeysierra** [comment]
Yes, backup requests and hedged requests are the same thing. Just different names.

**Comment by gelpenaddict** [comment]
Can you add a succeed on a Throwable?  So that it keeps retrying if there is no exception and succeeds when you get a particular exception.

**Comment by jhalterman** [comment]
@gelpenaddict Yea, by default throwables are handled (treated as failures) but you can override that via `handleIf`, and based your condition on a result, ignoring any failures (exception):

```java
retryPolicy.handleIf((result, failure) -> result == someBadValue)
```

**Comment by chadselph** [comment]
Ratelimit / bulkhead sounds really useful. For background refresh, what about an exponential backoff on failures?

**Comment by timothybasanov** [comment]
> Not a Contribution.

Not sure if I understand backup requests correctly, so I'm describing a behaviour that would be nice to have in my case:

Backup request is a bit different from a hedged one. It's more similar to an existing Fallback policy, but it starts its execution immediately, in parallel to the main request. Its results are _only_ used when the main request fails. It may also need some retries/circuit breakers/fallbacks configured.

**Comment by jhalterman** [comment]
@timothybasanov Can you open an issue for backup requests?
